src/backend/distributed/commands/database.c

```diff
@@ -107,3 +109,41 @@ get_database_owner(Oid db_oid)
 
 	return dba;
 }
+
+
+/*
+ * PreprocessGrantOnDatabaseStmt is executed before the statement is applied to the local
+ * postgres instance.
+ *
+ * In this stage we can prepare the commands that need to be run on all workers to grant
+ * on databases.
+ */
+List *
+PreprocessGrantOnDatabaseStmt(Node *node, const char *queryString,
+							  ProcessUtilityContext processUtilityContext)
+{
+	if (!ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	GrantStmt *stmt = castNode(GrantStmt, node);
+	Assert(stmt->objtype == OBJECT_DATABASE);
+
+	List *databaseList = stmt->objects;
+
+	if (list_length(databaseList) == 0)
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
```

Comment:
For our scenario, number 1 is applicable since we just don't want to deal with distributed objects but also we want to distribute non-distributed objects

Reply:
@JelteF do we need to call EnsureAllObjectDependenciesExistOnAllNodes  for each database? Since database is the first object to be created, We shouldn't need this IMO.

Reply:
I think if a user is granted permissions on a database that is distributed, then we should propagate those permissions and we should also ensure that the user exists on all nodes.

> Do we need to call EnsureAllObjectDependenciesExistOnAllNodes for each database? Since database is the first object to be created, We shouldn't need this IMO.

Databases are owned by users, so I think they can depend on users. However, for these grants we should only propagate if the database is already distributed. So I don't think ensuring the database exists is necessary. 

One edgecase is that if you specify multiple databases in the grant, and some are in pg_dist_object, but others are not, then what do you do? I think the best is to remove the ones that are not in pg_dist_object from the command. Propagating CREATE DATABASE is another task with some other difficulties.

Reply:
I'm confused in this description.
I checked all other grants and see that EnsureAllObjectDependenciesExistOnAllNodes is being executed just for the object for which grant is being executed 

<html>
<body>
<!--StartFragment-->

  | Dependencies
-- | --
Aggregate_Grant | Function
Any_GrantRole | Role
FDW_Grant | -
ForeignServer_Grant | -
Function_Grant | Function
Procedure_Grant | Function
Sequence_Grant | Sequence
Routine_Grant | Function
Schema_Grant | __
  |  

<!--EndFragment-->
</body>
</html>

Reply:
Alright, let's leave this as is for now then. We're getting to a point anyway where all roles should be in pg_dist_object, so the whole discussion is not super important anyway.